index.src.html

```diff
@@ -768,87 +790,66 @@ <h3 id="determine-requests-referrer">
     </li>
     <li>
-      If <var>request</var> is the result of following a redirect from
-      an HTTP <a>Redirection 3xx</a> <var>response</var>, then
-      let <var>referrerSource</var>
-      be <var>request</var>'s <code>referrer</code>, and
-      let <var>policy</var> be <var>response</var>'s associated referrer
-      policy. If <var>referrerSource</var> is null, then return <code>no
-      referrer</code> and abort these steps.
+      Let <var>environment</var> be <var>request</var>'s <code>client</code>.
     </li>
     <li>
-      Otherwise, if <var>request</var> is not the result of following a
-      redirect:
+      If <var>request</var>'s <code>referrer</code> is a URL, then let
+      <var>referrerSource</var> be <var>request</var>'s
+      <code>referrer</code>. Otherwise:
 
       <ol>
         <li>
-          Let <var>environment</var> be <var>request</var>'s <code>client</code>.
+          If <var>environment</var> is a <a>document environment</a>:
+
+          <ol>
+            <li>
+              Let <var>document</var> be the {{Document}} object of the
+              <a>active document</a> of the <a>browsing context</a> of
+              <var>environment</var>'s <a>responsible browsing context</a>.
+            </li>
+          </ol>
         </li>
         <li>
-          If <var>request</var> was initiated by an element with
-          a <code>referrerpolicy</code> content attribute,
-          let <var>policy</var> be the attribute's value. Otherwise,
-          let <var>policy</var> be the value
-          of <var>environment</var>'s <a>referrer policy</a>.
+          Otherwise, <var>environment</var> is a <a>worker environment</a>:
+
+          <ol>
+            <li>
+              Let <var>source</var> be the <a>API referrer source</a>
+              specified by the <a>incumbent settings object</a>.
+            </li>
+            <li>
+              If <var>source</var> is a URL, let <var>referrerSource</var>
+              be <var>source</var>, otherwise let <var>document</var> be
+              <var>source</var>.
+            </li>
+          </ol>
         </li>
         <li>
-          If <var>request</var>'s <code>referrer</code> is a URL, then let
-          <var>referrerSource</var> be <var>request</var>'s
-          <code>referrer</code>. Otherwise:
+          If <var>document</var> is set, execute the following steps:
 
           <ol>
             <li>
-              If <var>environment</var> is a <a>document environment</a>:
-
-              <ol>
-                <li>
-                  Let <var>document</var> be the {{Document}} object of the
-                  <a>active document</a> of the <a>browsing context</a> of
-                  <var>environment</var>'s <a>responsible browsing context</a>.
-                </li>
-              </ol>
+              If <var>document</var>'s <a>origin</a> is not a scheme/host/port
+              tuple (because, for example, it has been sandboxed into a unique
```

Comment:
You can use "opaque origin" here rather than negating scheme/host/port. See https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque.